pusher/crypto.py

```diff
@@ -0,0 +1,91 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import (
+    print_function,
+    unicode_literals,
+    absolute_import,
+    division)
+
+import hashlib
+import nacl
+import base64
+
+from pusher.util import (
+    ensure_text,
+    data_to_string)
+
+import nacl.secret
+import nacl.utils
+
+# The prefix any e2e channel must have
+ENCRYPTED_PREFIX = 'private-encrypted-'
+
+def is_encrypted_channel(channel):
+    """
+    is_encrypted_channel checks() if the channel is encrypted by verifying the prefix
+    """
+    if channel.startswith(ENCRYPTED_PREFIX):
+        return True
+    return False
+
+def is_encryption_master_key_valid(encryption_master_key):
+    """
+    is_encryption_master_key_valid() checks if the provided encryption_master_key is validby checking its length
+    """
+    if len(encryption_master_key) == 32:
+        return True
+
+    return False
+
+def generate_shared_secret(channel, encryption_master_key):
+    """
+    generate_shared_secret() takes a utf8-string
+    and returns the sha256 hash in utf8-string format
+    """
+
+    if encryption_master_key is not None:
+
+        encryption_master_key = ensure_text(encryption_master_key, "encryption_master_key")
+        channel = ensure_text(channel, "channel")
+
+        if is_encryption_master_key_valid(encryption_master_key):
+            # the key has to be 32 bits long
+            return hashlib.sha256((channel + encryption_master_key).encode('utf-8')).hexdigest()[:32]
+
+    raise ValueError("Provided encryption_master_key is not 32 char long")
+
+def encrypt(channel, data, encryption_master_key, nonce):
+    """
+    encrypt() encrypts the provided payload specified in the 'data' parameter
+    """
+    shared_secret = generate_shared_secret(channel, encryption_master_key)
+    # the box setup to seal/unseal data payload
+    box = nacl.secret.SecretBox(shared_secret.encode("utf-8"))
+
+    """ this is here for reference the nonce
+        it is currently passed to the function
+        in order to allow testablity
+
+        # generate the nonce with nacl
+        #nonce = nacl.utils.random(nacl.secret.SecretBox.NONCE_SIZE)
+    """
+
+    nonce = ensure_text(nonce, "nonce")
```

Comment:
Yes, we should assume the nonce, either generated internally or provided, to be `bytes`.
This has a small caveat since bytes are not represented by the same types in Python 2.x and Python 3.x ( more here: https://pythonhosted.org/six/#six.binary_type and here: https://hacktec.gitbooks.io/effective-python/content/en/Chapter1/Item3.html ) and we should keep this in consideration for the edit.